The `#[diagnostic]` attribute namespace (#3368)
rust-lang/rfcs: text/3366-diagnostic-attribute-namespace.md

```

I expect the new attributes to be documented on the existing [Diagnostics](https://doc.rust-lang.org/reference/attributes/diagnostics.html) attributes page on the rust reference similar to existing attributes like for example `#[deprecated]`

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## The `#[diagnostic]` attribute namespace

This RFC proposes to introduce a new `#[diagnostic]` attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific error messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes.

Any attribute in this namespace may:

* Hint the compiler to emit a specific error message in a specific situation

Any attribute in this namespace is not allowed to:

* Change the result of the compilation, which means applying such an attribute should never cause an error as long as they are syntactically valid

> 

Reviewer: Does this promise that "new" syntax in the attribute will compile (but perhaps do nothing) backwards compatibly? 

For example,
```rust
#[cfg_attr(FALSE, doc = include!(""))]
struct Blah;
```
...does nothing, in any compiler version, but does not compile below Rust 1.54. (it gives an unstable error from 1.50, and a syntax error before that).

Personally I don't care much about error messages on non-latest rust, and would happily use new attributes and arguments as they become available. But I would dislike new syntax incompatible with my msrv, because I would have to resort to `cfg_attr` tricks to disable it.

In other words, if this namespace stabilizes in a given Rust version, is it guaranteed that it will remain syntactically valid for that version?

Reviewer 2: I believe what's proposed is only forwards compatibility and not backwards compatibility: If `diagnostics::on_type_error` is stabilized in 1.N, using it in 1.(N-1) will result in a compilation error. The same for any extensions of the API of `diagnostics::on_unimplemented`. I would *personally* prefer an approach that is more lenient, *but* even if the language won't do that, once the feature is stabilized a third party crate can act as the glue to accomplish that, by doing version checking and translating from the provided proc macro to the supported API surface (including doing nothing).

Reviewer 2: @weiznich I'm happy enough with that solution because we can then provide a crate for those who care about lower MSRV, in the same spirit as `standback`.

The compiler is allowed to:

* Ignore the provided hints
* Use only parts of the provided hints, for example for the proposed `#[diagnostic::on_unimplemented]` only use the `message` option

> * Change this behaviour at any time

Reviewer: I am happy with api evolution, but I'd like to be aware if/when this happens. I don't expect it to actually change much (after all, `rustc_on_unimplemented` has barely changed backwards compatibly), but I am a bit afraid of having a lot of these attributes in my codebase, only for future me to have that "oh this hasn't actually worked for a long time" moment.

Perhaps a lint diagnostic should be emitted for that?

Reviewer 2: `rustc_on_unimplemented` hasn't ever broken backwards compatibility, somehow 😅

 > Perhaps a lint diagnostic should be emitted for that?

+1

Author: I personally would consider that a quality of implementation issue. Which means that's something where a specific compiler implementation can offer a lint for. Other than that: At least for my own use-cases I will likely have tests based on `trybuild` or something else. These will show whether the attributes are ignored or not.

## The `#[diagnostic::on_unimplemented]` attribute

This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. Implementing the later part is optional as outlined in the previous section. Any syntax not explicitly described in this section must be rejected as invalid according to the general rules of the `#[diagnostic]` attribute namespace.

```rust
#[diagnostic::on_unimplemented(
    message="message",
    label="label",
    note="note",
)]
trait MyIterator<A> {
    fn next(&mut self) -> A;
}
```

Each of the options `message`, `label` and `note` are optional. They are separated by comma. The tailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. The error message can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the actual type name.

> In addition the `on_unimplemented` attribute provides mechanisms to specify for which exact types a certain message should be emitted via an `if()` option. It accepts a set of filter options. A filter option consists on the generic parameter name from the trait definition and a type path against which the parameter should be checked. These type path could either be a fully qualified path or refer to any type in the current scope. As a special generic parameter name `Self` is added to refer to the `Self` type of the trait implementation. A filter option evaluates to `true` if the corresponding generic parameter in the trait definition matches the specified type. The provided `message`/`note`/`label` options are only emitted if the filter operation evaluates to `true`.

Reviewer 2: Is there a reason to change `on` to `if` from `rustc_on_unimplemented`? I don't care either way, I'm just intrigued at the rationale. `on` is already used in other attributes IIRC.

Also, allowing `Self` is something I'd want and is easy to add, but might have pushback from the lang team.

Author: I've went for `if` instead of `on` as I also moved the `message`/`notice`/`label` option out of the filter clause. Instead each attribute now has an optional filter in the `if` clause that determines whether a `message`/… should be emitted. I consider the multiple attributes approach to be easier to use in conjunction with something like `#[cfg_attr]` especially around filters containing types behind a feature flag. So `if` feels for fitting for my non-native speaker mind. If `on` is a better fit I'm happy to use `on` instead.

The `any` and `all` option allows to combine multiple filter options. The `any` option matches if one of the supplied filter options evaluates to `true`, the `all` option requires that all supplied filter options evaluate to true. `not` allows to negate a given filter option. It evaluates to `true` if the inner filter option evaluates to `false`. These options can be nested to construct complex filters.

> The `on_unimplemented` attribute can be applied multiple times to the same trait definition. Multiple attributes are evaluated in order. The first matching instance for each of the `message`/`label`/`note` options is emitted.

Reviewer 2: I feel like the bulk of the `on_unimplemented` behavior should be on its own RFC, because there are open questions about it. What happens when you have multiple cases that match and the attribute writer doesn't notice? Is that detected by the compiler and emits an error?

Reviewer 3: Yeah, the discussion of `on_implemented` is so involved that it looks like it's proposed to be stabilized in this RFC. If it isn't, it needs to be trimmed down from irrelevant details and more prominently marked as just an example.

Author: I've opted to put everything into one RFC for now as the proposed `#[diagnostic]` attribute namespace is only useful with actual attributes that can be used by someone. So yes: This RFC proposes both, the namespace and the `#[on_unimplemented]` attribute. If it turns out that there is no conses around how the `filter` rules should be designed I would propose to only specify the `#[on_unimplemented]` attribute without the filtering functionality. So just the `message`/`label`/`note` options. I would expect that it is much easier to get these options approved.

Reviewer 2: I think that's a reasonable approach: land `#[diagnostics::on_unimplemented(note="", message="", label="")]` with this RFC and a subsequent RFC for the evolution of `on_unimplemented`. This would also (I think) help t-lang to experience what evolving these would look like in the future and catch any potential problems or concerns *with a concrete example*.

Author: I would nevertheless wait for t-lang before removing the relevant RFC section.This would at least give some initial feedback on what needs to be addressed in a future RFC.

```rust
#[diagnostic::on_unimplemented(
    if(Self = std::string::String),
    note = "That's only emitted if Self == std::string::String"
)]
#[diagnostic::on_unimplemented(
    if(A = String), // Refers to whatever `String` is in the current scope
    note = "That's only emitted if A == String",
)]
#[diagnostic::on_unimplemented(
    if(any(A = i32, Self = i32)),
    note = "That's emitted if A or Self is a i32",

> )]

Reviewer: Should this check whether a type has the *name* String or whether it *is* String? `#[rustc_on_unimplemented]` [blindly checks the name of the type](https://play.integer32.com/?version=stable&mode=debug&edition=2021&gist=a24ff48dc2174de7692484b622c4f6c2), but "Refers to whatever `String` is in the current scope" sounds like you want the latter.

Reviewer 2: I would anticipate that `on_unimplemented` would receive the `DefId` of whatever path was passed in here, so it would be following regular visibility rules.

Author: I propose to use the regular visibility rules here instead of just matching the name as string as that's currently done. I feel that aligns the proposed attribute more with other parts of rust.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This proposal tries to improve error messages generated by rustc. It would give crate authors a tool to influence what error message is emitted in a certain situation, as they might sometimes want to provide specific details on certain error conditions. Not implementing this proposal would result in the current status quo. Currently the compiler always shows a "general" error message, even if it would be helpful to show additional details.

There are alternative designs for the proposed `on_unimplemented` attribute:

* The `on()` based filtering might be replaceable by placing the attribute on negative trait impls. This would turn a filter like
```rust
#[diagnostic::on_unimplemented(
    on(Self = `String`, message = "Strings do not implement `IntoIterator` directly")
)]
trait IntoIterator {}
```

into the following negative trait impl:
```rust
#[diagnostic::on_unimplemented(message = "Strings do not implement `IntoIterator` directly")]
impl !IntoIterator for String {}
```

> This would simplify the syntax of the proposed attribute, but in turn block the implementation of type based filtering on the stabilization of `negative_impls`. On the other hand it would likely simplify writing more complex filters, that match only a certain generic set of types and it would prevent "duplicating" the filter-logic as this reuses the exiting trait system. The large disadvantage of this approach is that it couples error messages to the crates public API. Removing a negative trait impl is a breaking change, removing a `#[on_unimplemented]` attribute is only a change in the emitted compiler error.

Reviewer 2: You need to call out that the simplification of complex filtering logic also requires some level of specialization support. This would be fine because even if the current impl is unsound it doesn't affect this use case, but it might close doors and block future specialization behavior changes.

Author: That's a good point. Will add that :+1:

Reviewer 2: Another thing is that doing the attribute approach now *doesn't* preclude us from implementing a more integrated system in the future.

> * Allow `#[diagnostic::on_unimplemented]` to be placed on types instead of traits. This would allow third party crates to customize the error messages emitted for unsatisfied trait bounds with out of crate traits. This feels more like an extension tho the proposed attribute.

Reviewer 2: I don't see this as an alternative but more as future work.

Reviewer: I'd like to have both (eventually). Current `#[rustc_on_unimplemented]` works well mostly because std is mostly defining the traits, not using traits defined by others.

Author: Will move that point to future possibilities :+1:

# Prior art
[prior-art]: #prior-art

* [rustc_on_unimplemented](https://rustc-dev-guide.rust-lang.org/diagnostics.html#rustc_on_unimplemented) already provides the described functionality as rustc internal attribute. It is used for improving error messages for various standard library API's. [This repo](https://github.com/weiznich/rust-foundation-community-grant/) contains several examples on how this attribute can be used in external crates to improve their error messages.

> * [GHC](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/type_errors.html) provides an Haskell language extension for specifying custom compile time errors

Reviewer 4: ```suggestion
* [GHC](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/type_errors.html) provides a Haskell mechanism for specifying custom compile time errors
```
`TypeError` isn't really a language extension, it's a builtin type family (a type-level function) that uses several GHC extensions.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This proposal tries to improve error messages generated by rustc. It would give crate authors a tool to influence what error message is emitted in a certain situation, as they might sometimes want to provide specific details on certain error conditions. Not implementing this proposal would result in the current status quo. Currently the compiler always shows a "general" error message, even if it would be helpful to show additional details.

There are alternatives for the naming of the `#[diagnostic]` namespace:

* Use a `#[rustc]` namespace for these attributes. This would signifies that these are rustc specific extensions. We likely want to encourage other rust implementations to utilise these information as well, therefore a more general attribute namespace seems to be a better solution.

There are alternative designs for the proposed `on_unimplemented` attribute:

> * The `on()` based filtering might be replaceable by placing the attribute on negative trait impls. This would turn a filter like

Reviewer 5: `on()` has not been defined in the above description of `on_unimplemented`? it seems like the above description changed that to `of`.

Author: Good catch. That should be address in https://github.com/rust-lang/rfcs/pull/3368/commits/cb8f2fc5aec3b4f2d35a6f017fd7e9e59eb91fa7

[reference-level-explanation]: #reference-level-explanation

## The `#[diagnostic]` attribute namespace

This RFC proposes to introduce a new built-in `#[diagnostic]` tool attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific diagnostic messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes. This opens the possibility to delegate the design of specific attributes to other teams later on.

Attributes in this namespace are generally expected to be formed like:
```rust
#[diagnostic::attribute(option)]
```
where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. 

Any attribute in this namespace may:

* Hint the compiler to emit a specific diagnostic message in a specific situation
* Only affect the messages emitted by a compiler 

Any attribute in this namespace is not allowed to:

> * Change the result of the compilation, which means applying such an attribute should never cause a compilation error as long as they are syntactically valid

Reviewer 6: Should this say “applying _or removing_ such an attribute should never cause a compilation error”?

Author: I would argue that this is already somewhat implied by the first part of the sentence (cannot change the result of the compilation). That written, it might be meaningful to add it explicitly to the text as well.

* Pass-through information from the source of the diagnostic in a way that users can rely on. E.g. Such an attribute should not allow users to keep the compilation successful and dump information about `extern` blocks to generate C header files

The compiler is allowed to:

* Ignore the hints provided by:
    + A specific attribute
    + A specific option 
* Change the support for a specific attribute or option at any time

The compiler must not:

> * Change the semantic of an attribute or option

Author: Its for supposed to prevent that compiler starts at to emit what's passed as `message` option as `label` or that the compiler starts at one point to use the `on_unimplemeted` attribute for type mismatches (however that would be possible...)
Maybe I should add these cases as example to the RFC as well?

* Emit an hard error on malformed attribute

The compiler should:

* Emit implement a warn-by-default lint for unrecognised attributes or options 

Adding a new attribute or option to the `#[diagnostic]` namespace is for now a decision of the language team. The language team can delegate these decisions partially or completely to a different team without requiring a new RFC. 

## The `#[diagnostic::on_unimplemented]` attribute

This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. The specification of this attribute is partially provided as example and motivation for the `#[diagnostic]` attribute namespace. In addition it is provided to give this RFC a concrete use, such that we not only define an empty attribute namespace.

```rust
#[diagnostic::on_unimplemented(
    message="message",
    label="label",
    note="note",
)]
trait MyIterator<A> {
    fn next(&mut self) -> A;
}
```

> Each of the options `message`, `label` and `note` are optional. They are separated by comma. The trailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. These options can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the compiler with the actual type name.

Reviewer 7: I understand the idea of *replacing* the message and the label, in the sense that I think there is always a message, and there is one label (potentially *at most one* label? Can the label be omitted? Is this inserted in its stead in that case?). So replacing each of those is the sensible option here.

However, a given diagnostic can have zero, one, or more than one notes supplied, right?. What is *replaced* for the "more than one" case, here? And would it be better to just accumulate this note onto the set, or are we indeed better off sticking with replacement here?

Reviewer 2: What the best approach is depends entirely on the situation, sadly (which tells me it should be configurable somehow). Currently `rustc_on_unimplemented` will replace these but add a note with the original message.

I can see for example errors that would be very verbose where the underlying problem is very simple, so as a crate author I'd like to be able to almost entirely clear the diagnostic and clearly state "this is wrong and this is how you fix it", avoiding that messaging being lost in the noise. That being said there would have to be a consideration of whether this kind of "obscuring" of the emitted error could constitute a problem if misused. All this being said, I believe these are specific questions to be had around the stabilization of `#[diagnostic::on_unimplemented]` and not the acceptance of `#[diagnostic::*]` as a construct.

- Feature Name: The `#[diagnostic]` attribute namespace
- Start Date: 2023-01-06
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> This RFC proposed to add a stable `#[diagnostic]` attribute namespace, which contains attributes to influence error messages emitted by the compiler. In addition it proposed to add a `#[diagnostic::on_unimplemented]` attribute to influence error messages emitted by unsatisfied traits bounds.

Reviewer 8: ```suggestion
This RFC proposed to add a stable `#[diagnostic]` attribute namespace, which contains attributes to influence error messages emitted by the compiler. Attributes in this namespace are versioned and crates that use the `#[diagnostic]` attributes must declare a version at the root level (e.g., `#[diagnostic::v1]`). In addition it proposes to add a `#[diagnostic::on_unimplemented]` attribute to influence error messages emitted by unsatisfied traits bounds.
```

Author: Good catch, that's something I've missed after updating the RFC to incorporate a version mechanism. Should I push an update now or should I wait for the [proposed design meeting](https://hackmd.io/e2fvDhvFS_-muKmxOezX_A#%E2%80%9CThe-diagnostic-attribute-namespace%E2%80%9D-rfcs3368) to happen?

14 |     iterate_chars(&[1, 2, 3][..]);
   |     ^^^^^^^^^^^^^ label
   |
   = note: note
   = help: the trait `MyIterator<char>` is not implemented for `&[{integer}]`
   = note: required by `iterate_chars`
```

I expect the new attributes to be documented on the existing [Diagnostics](https://doc.rust-lang.org/reference/attributes/diagnostics.html) attributes page on the rust reference similar to existing attributes like for example `#[deprecated]`

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## The `#[diagnostic]` attribute namespace

This RFC proposes to introduce a new `#[diagnostic]` attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific error messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes.

Attributes in this namespace are generally expected to be formed like:
```rust
#[diagnostic::attribute(option)]
```

> where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. Attributes in this namespace are versioned. Any incompatible change to an existing stable attribute in this namespace requires bumping the version of the diagnostic attribute namespace. Crates using these attributes must specify the expected version of the diagnostic namespace via a `#![diagnostic::v{version_number}]` crate top-level attribute. If multiple such attributes are present the compiler will choose the highest version number and ignore the rest. The compiler is encouraged to provide a lint for outdated versions of the diagnostic attribute namespace. This mechanism exists so that a third party crate can provide a shim implementation of the corresponding attributes to translate between different versions of the diagnostic attribute namespace on different rust versions.

Reviewer: Today I was thinking about how to design a macro api, and I realized a (niche) downside of this versioning approach. It won't work (out of the box) when I want to insert it in my user's code and I own neither the trait nor the struct.

For example, say an user writes the following code:
```rust
#[make_secure] // my macro!
struct Password(String);
```

I'd really like to make this expand to:

```rust
struct Password(String);

// imagine a world where negative_impls is stable
#[diagnostic::on_unimplemented(message = "secure structs may not be printed")]
impl !Debug for Password {}
```

Is there a good workaround for this? Or will this cause the compiler to throw a "please enable diagnostics version 12 for diagnostics on negative impls" error to my user?

Reviewer 2: I think that *per attr* versioning selection (a convention of accepting a specific field?) would need to be provided to cater to attribute proc macros, as you correctly point out :-/

Reviewer 9: could hygiene pick the macro crate's attribute version instead of the user crate's version? Regular macros would have the same issue.

> where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. Attributes in this namespace are versioned. Any incompatible change to an existing stable attribute in this namespace requires bumping the version of the diagnostic attribute namespace. Crates using these attributes must specify the expected version of the diagnostic namespace via a `#![diagnostic::v{version_number}]` crate top-level attribute. If multiple such attributes are present the compiler will choose the highest version number and ignore the rest. The compiler is encouraged to provide a lint for outdated versions of the diagnostic attribute namespace. This mechanism exists so that a third party crate can provide a shim implementation of the corresponding attributes to translate between different versions of the diagnostic attribute namespace on different rust versions.

Reviewer 10: I don't think I'm comfortable with adding a new versioning scheme for a subset of the language. Or at least, I don't understand the motivation at all. Why can't we use editions to evolve the attributes?

> This mechanism exists so that a third party crate can provide a shim implementation of the corresponding attributes to translate between different versions of the diagnostic attribute namespace on different rust versions.

What purpose does this serve? Supporting newer versions on old compilers? How would it work?

Reviewer 9: > Why can't we use editions to evolve the attributes?

because we want to remove old versions from the compiler entirely. Editions must be supported until the heat death of the universe.

> What purpose does this serve? Supporting newer versions on old compilers? How would it work?

a crates.io crate could check the compiler version and change what attribute they generate, even if the crate's input attribute syntax is stable across compiler versions.

Author: For context: This versioning scheme was proposed by @joshtriplett. I personally would be fine with specifying that the compiler is not allowed to change the semantic of a certain attribute/option + it is allowed to ignore a certain attribute/option. If necessary that alone would allow to replace the a certain option by another one by just choosing a differently named option.

Any attribute in this namespace may:

* Hint the compiler to emit a specific error message in a specific situation
* Only affect the messages emitted by a compiler in case of a failed compilation

Any attribute in this namespace is not allowed to:

* Change the result of the compilation, which means applying such an attribute should never cause an error as long as they are syntactically valid
* pass-through information from the source of the diagnostic in a way that users can rely on. E.g. Such an attribute should not allow users to keep the compilation successful and dump information about `extern` blocks to generate C header files

The compiler is allowed to:

* Ignore the hints provided by:
    + A specific attribute
    + A specific option 
* Change the set of supported attributes and options at any time
* Provide warning lints for malformed or otherwise not accepted attributes 

The compiler must not:

* Change the semantic of an attribute or option, these kind of change requires a bump of the diagnostic attribute space version number
* Emit an hard error on malformed attribute, although emitting lints is fine

> Adding a new attribute or option to the `#[diagnostic]` namespace is a decision of the compiler team. The envisioned procedure for this is as following:

Reviewer 9: > Put another way, is the justification for bypassing T-lang approval essentially that: it doesn't affect semantics of programs, including whether they compile, and that alternate compilers can ignore it?

rustc can ignore them, they are effectively a new kind of comments from the lang perspective, even if we as compiler developers will want to give users a reliable experience.

> Adding a new attribute or option to the `#[diagnostic]` namespace is a decision of the compiler team. The envisioned procedure for this is as following:

Author: The original proposal (before opening this RFC) was to introduce a `#[rustc::]` attribute namespace. That would make it much clearer that this these attributes are compiler specific. Some members of the compiler team and the language team raised the valid concern that these specific attributes around diagnostics might be relevant for other compilers as well. So a different named attribute namespace seems to be more fitting.

Now this RFC tries to specify the namespace in such a way that supporting anything in that namespace should be optional for other compilers. So if these attributes do not fit their model, they are free to ignore these attributes (or ignore certain options only). Otherwise the main argument for "bypassing" approval from the language team is to prevent spamming the team with otherwise trivial changes. Again that was something that was proposed by @joshtriplett .

1. The new attribute lands on nightly without a feature gate. It gets ignored on stable/beta until its explicitly activated there.
2. The time as nightly only attribute is used to verify that:
    * The attribute solves a real problem
    * The implemented syntax matches the expectations
3. To enable a certain attribute in stable release the following steps are required:
    * The attribute is documented in the rust reference
    * T-lang is informed
    * T-compiler performs a FCP, where a simple majority (>50%) accepts the attribute
    

## The `#[diagnostic::on_unimplemented]` attribute

This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. The specification of this attribute is partially provided as example and motivation for the `#[diagnostic]` attribute namespace. In addition it is provided to give this RFC a 

> 

Author: Seems like I just stopped writing in the middle of that sentence 🤦 
I've pushed a change to finish that sentence.

The compiler is allowed to:

* Ignore the hints provided by:
    + A specific attribute
    + A specific option 
* Change the set of supported attributes and options at any time
* Provide warning lints for malformed or otherwise not accepted attributes 

The compiler must not:

* Change the semantic of an attribute or option, these kind of change requires a bump of the diagnostic attribute space version number
* Emit an hard error on malformed attribute, although emitting lints is fine

Adding a new attribute or option to the `#[diagnostic]` namespace is a decision of the compiler team. The envisioned procedure for this is as following:

1. The new attribute lands on nightly without a feature gate. It gets ignored on stable/beta until its explicitly activated there.
2. The time as nightly only attribute is used to verify that:
    * The attribute solves a real problem
    * The implemented syntax matches the expectations
3. To enable a certain attribute in stable release the following steps are required:
    * The attribute is documented in the rust reference
    * T-lang is informed

>     * T-compiler performs a FCP, where a simple majority (>50%) accepts the attribute

Reviewer 9: Leave a comment saying that the usual concern rules still apply, everyone can raise a concern.
